prog.go

```diff
@@ -35,14 +35,54 @@ const (
 // verifier log.
 const DefaultVerifierLogSize = 64 * 1024
 
+// LogLevel controls the verbosity of the kernel's eBPF program verifier.
+// These constants can be used for the ProgramOptions.LogLevel field.
+type LogLevel uint32
+
+const (
+	// Print verifier state at branch points.
+	LogLevelBranch LogLevel = iota + 1
+
+	// Print verifier state for every instruction.
+	// Available since Linux v5.2.
+	LogLevelInstruction
+
+	// Print verifier errors and stats at the end of the verification process.
+	// Available since Linux v5.2.
+	LogLevelStats
+
+	// Disable the verifier log.
+	LogLevelDisabled = math.MaxUint32
+)
+
 // ProgramOptions control loading a program into the kernel.
 type ProgramOptions struct {
-	// Controls the detail emitted by the kernel verifier. Set to non-zero
-	// to enable logging.
-	LogLevel uint32
-	// Controls the output buffer size for the verifier. Defaults to
-	// DefaultVerifierLogSize.
-	LogSize int
+	// Bitmap controlling the detail emitted by the kernel's eBPF verifier log.
+	// LogLevel-type values can be ORed together to request specific kinds of
+	// verifier output. See the documentation on [ebpf.LogLevel] for details.
+	//
+	//  opts.LogLevel = (ebpf.LogLevelBranch | ebpf.LogLevelStats)
+	//
+	// If left to its default value, the program will first be loaded without
+	// verifier output enabled. Upon error, the program load will be repeated
+	// with LogLevelBranch and the given (or default) LogSize value.
+	//
+	// Setting this to a value other than LogLevelDisabled will unconditionally
+	// enable the verifier log. This will always allocate an output buffer, but
+	// will result in only a single attempt at loading the program.
+	LogLevel LogLevel
+
+	// Controls the output buffer size for the verifier log, in bytes. See the
+	// documentation on ProgramOptions.LogLevel for details about how this value
+	// is used.
+	//
+	// If this value is set too low to fit the verifier log, the resulting
+	// [ebpf.VerifierError]'s Truncated flag will be true, and the error string
+	// will also contain a hint to that effect.
+	//
+	// Defaults to DefaultVerifierLogSize.
+	LogSize uint64
```

Comment:
Is uint64 really appropriate here? The go allocator is limited to int anyways (?) so there is no need to break compat.

Reply:
If we allow negative values, we need to sanitize, or the call to `make([]byte, opts.LogSize)` will panic. Previously, there was a check on `LogSize >= 0`, but why allow negative values to begin with?

Reply:
I'm not saying that negative size makes sense, I'm saying that `int` is the natural choice for allocations. I'm fine panicking with negative size since that's what the runtime does.

Another thought that shows why we should stick to what the runtime does. why use uint64 not uint? We can't allocate that much memory on 32 bit anyways.

(One caveat of course: we used hacky negative size to disable the log, but never documented that. We'd have to hope that nobody outside of the lib uses that.)

Reply:
> ... we used hacky negative size to disable the log, but never documented that. We'd have to hope that nobody outside of the lib uses that.

I'll change this back to an int and throw an error when it's negative, including a pointer to the new bool. Might save a user some time debugging after upgrading. As long as there's a way to disable the log and it's a quick fix, there should not be any complaints.

This should also be marked as a breaking change in the relnotes.